ciba doc (#3762)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/configure-ciba/main/index.md

---
title: Transactional verification using CIBA

> excerpt: Learn how to use transactional verification with CIBA and a custom authenticator

Author: ```suggestion
excerpt: How to implement transactional verification with CIBA and a custom authenticator
```

@startuml
skinparam monochrome true
participant "OIDC client" as client
participant "Authorization server" as as
participant "Authentication device" as device

autonumber "<b>#."
client -> as: Makes POST request to authorization server
as -> client: Sends backchannel authentication response with `auth_req_id`
as -> device: Makes background request for user consent/authorization
client -> as: Begins polling with `auth_req_id` to `/token` for tokens
as -> client: Sends authorization error
device -> as: Returns user consent/authorization
client -> as: Polls with `auth_req_id` to `/token` for tokens
as -> client: Returns requested tokens
@enduml

-->

1. The OIDC application on the consumption device triggers the CIBA authentication flow by sending a POST request to the authorization server’s backchannel authentication endpoint (`/oauth2/v1/bc/authorize`).

    To identify the user that additional authentication is being requested for, the request must include either a previously issued ID token (obtained during the user’s initial authorization) as an `id_token_hint` or the user’s username (email address) as a `login_hint`.

>     > **Note:** The `id_token_hint` is used in cases where the user is signed in to the application running on the consumption device, and the user is authenticated by the same authorization server. The user is also in control of both the consumption and authentication devices.

Author: @venkatviswanathan-okta - is this ok?

Reviewer: Yes

    > **Note:** You can leave the rest of the default values, as they work with this guide for testing purposes.

9. In the **Assignments** section, select **Allow everyone in your organization to access**.
10. Click **Save**. The settings page for the app integration appears, showing the **General** tab.
11. Make note of the **Client ID** in the **Client Credentials** section and the client secret in the **CLIENT SECRETS** section. You need this information for the [Test the CIBA flow](#test-the-ciba-flow) section.

#### Use the API

You can also use the [Apps API](/docs/reference/api/apps/#add-oauth-2-0-client-application) to create the OAuth 2.0 client app with the CIBA grant type. Use the following parameters in the request:

**Supported parameters**

| Parameter                    | Value                                                                        |
| :--------------------------- | :--------------------------------------------------------------------------- |
| `grant_types`                | `urn:openid:params:grant-type:ciba`     |

> | `backchannel_authentication_request_signing_alg` | (Optional) The signing algorithm for CIBA signed requests using JWT. If this value isn't set and a JWT signed request is sent, the request fails. |

Reviewer 2: ```suggestion
| `backchannel_authentication_request_signing_alg` | (Optional) The signing algorithm for CIBA-signed requests using JWT. If this value isn't set and a JWT-signed request is sent, the request fails. |
```

Reviewer 2: if you think it's too hyphen heavy, np.

**Example request**

```json
{
    "name": "oidc_client",
    "label": "CIBA Client",
    "signOnMode": "OPENID_CONNECT",
    "credentials": {
      "oauthClient": {
        "token_endpoint_auth_method": "client_secret_post"
      }
    },
    "settings": {
      "oauthClient": {
        "client_uri": "http://localhost:8080",
        "logo_uri": null,
        "redirect_uris": [
          "https://example.com/oauth2/callback",
          "myapp://callback"
        ],
        "response_types": [
          "token",
          "id_token",
          "code"
        ],
        "grant_types": [
          "implicit",
          "authorization_code",
          "urn:openid:params:grant-type:ciba"
        ],
        "application_type": "web",
        "backchannel_custom_authenticator_id" : "aut5ur07m58W2sQD01d7"
      }
    }
}
```

> **Note:** You can also use the Apps API to [update an OAuth 2.0 client app to use CIBA](/docs/reference/api/apps/#update-application). See the **Supported parameters** table for the parameters to use.

### Configure an authorization server access policy to use CIBA

> You can use either the Okta Org Authorization Server or a custom authorization server with CIBA. In this use case example, we use the `default` custom authorization server.

Reviewer 2: ```suggestion
You can use either the Okta Org Authorization Server or a custom authorization server with CIBA. In this example, use the `default` custom authorization server.
```

1. In the Admin Console, go to **Security** > **API**.

> 2. On the **Authorization Servers** tab, select the edit icon to the right of the authorization server that you want to use. In this use case example, we are using the `default` custom authorization server.

Reviewer 2: ```suggestion
2. On the **Authorization Servers** tab, select the edit icon to the right of the authorization server that you want to use. In this example, use the `default` custom authorization server.
```

3. Select **Access Policies**.
4. Select the edit icon for the Default Policy Rule.

> 5. Select **Client-initiated backchannel authentication (CIBA)** as a grant type and click **Update Rule**.

Reviewer 2: ```suggestion
5. Select **Client-initiated backchannel authentication (CIBA)** as a grant type and click **Update rule**.
```

## Enroll a custom authenticator to use CIBA

> Use the Devices SDK and your app to enroll a custom authenticator for the test user. In this use case example, we are using the Okta Authenticator Sample App and the xCode simulator to enroll the authenticator. See the **What you need** section if you haven’t set up the sample app.

Reviewer 2: ```suggestion
Use the Devices SDK and your app to enroll a custom authenticator for the test user. In this example, use the Okta Authenticator Sample App and the Xcode simulator to enroll the authenticator. See the **What you need** section if you haven’t set up the sample app.
```

1. In the **PushSampleApp** directory of the Okta Devices SDK, run the Okta Authenticator Sample App and then open the mobile app in the simulator.

> 2. Click **Sign In**, enter the test user’s credentials, and click **Sign In**.

Reviewer 2: ```suggestion
2. Click **Sign In**. Enter the test user’s credentials, and click **Sign In**.
```

Author: I want to keep this as is.

3. Complete any additional two-factor authentication required and click **Verify**.

> 4. On the Sign in securely using push notifications screen, click **Set up**.

Reviewer 2: Might be overkill, but should "Sign in securely using push notifications" be in bold. It's a mouthful without context.

5. On the Security Settings screen, slide to enable **Sign in with push notifications**.
6. Click **Ok** at the confirmation dialog. Additional security setting options then appear, including **Enable CIBA transactions**.
7. Slide to select **Enable CIBA transactions**.<br>
   The sample app is set up to include the **Enable CIBA transactions** option by default for the user to enable CIBA themselves in the mobile app. However, you can implement CIBA in your app any way that you want, for example, create an enrollment flow that turns CIBA on by default, making it transparent to users.
8. Click **Ok** at the success dialog and close the Security Settings screen.

> 9. Leave your test user signed in to the authentication device (the Magenta Bank app on the xCode simulator).

Reviewer 2: ```suggestion
9. Leave your test user signed in to the authentication device (the Magenta Bank app on the Xcode simulator).
```

  <!-- **Note:** See the [MyAccount API](/docs/reference/api/myaccount/) for examples of [enrolling](?link to new api operation?) and [updating](?link to new api operation?) a custom app authenticator to use CIBA by adding `CIBA` to the `transactionTypes` array. -->

## Test the CIBA flow

This section walks you through how to test the CIBA authentication flow using the Okta Authenticator Sample App and the custom authenticator that you previously set up.

### Base64-encode the client ID and secret

> The call to the `/bc/authorize` endpoint and to the `/token` endpoint requires authentication. In this case, it’s a Basic Auth digest of the client ID and secret of the OIDC client that you created in [an earlier section of the guide](#create-an-oidc-client-integration-and-enable-CIBA). You can find them on the client’s **General** tab in the **Client Credentials** section.

Reviewer 2: Check this link/anchor. It wasn't working in my preview. #create-an-oidc-client-integration-and-enable-ciba is the link in my preview (lower-case ciba)

To use a Basic Authorization header in both the `/bc/authorize` and `/token` requests, Base64-encode the string and set it in the Authorization header:

```sh
   Authorization: Basic ${Base64(${clientId}:${clientSecret})}
```

### Initiate the flow

The CIBA challenge request is sent by an OIDC client to an Okta authorization server. This request can include the following parameters:

**Supported parameters**

| Parameter                    | Value                                                                        |
| :--------------------------- | :--------------------------------------------------------------------------- |
| `binding_message`| (Optional) A human-readable message that appears on the authentication device to identify the transaction. |

> | `id_token_hint` | An ID token that was issued during initial user authentication. The token is passed back as a hint to identify the user for whom additional authentication is being requested. You can specify either `login_hint` or `id_token_hint` in the authentication request, not both. This parameter isn’t used in this use case example. |

Reviewer 2: ```suggestion
| `id_token_hint` | An ID token that was issued during initial user authentication. The token is passed back as a hint to identify the user for whom additional authentication is being requested. You can specify either `login_hint` or `id_token_hint` in the authentication request, not both. This parameter isn’t used in this example. |
```

| `login_hint` | Information that identifies the user for whom authentication is being requested. This is typically the user’s email address. You can specify either `login_hint` or `id_token_hint` in the authentication request, not both. |
| `request`| If you are using `private_key_jwt` as the [token endpoint authentication method](/docs/reference/api/apps/#credentials), this is the JWT created by the client that enables you to pass requests as a single, self-contained parameter. See [Parameter details](/docs/reference/api/oidc/#client-initiated-backchannel-authentication-parameter-details). |
| `request_expiry`| (Optional) A positive integer that allows the client to request how long (in seconds) the authentication request is valid. This value is returned in the response from the authorization server as the `expires_in` parameter. If your request doesn’t contain this parameter, the default time that the request is valid is 300 seconds. |
| `scope` | The `openid` scope is required for authentication requests. You can also include other [scopes](/docs/reference/api/oidc/#access-token-scopes-and-claims).|

> > **Note:** For additional request examples using different parameters as well as a public key/private key [client authentication](/docs/reference/api/oidc/#client-authentication-methods) example, see [CIBA request examples](/docs/reference/api/oidc/#ciba-request-examples).

Reviewer 2: I didn't see this reference, ciba request examples. Is that in another PR?

Author: Yes, this is in Mymoona's CIBA PR #3709

| `grant_type`| `urn:openid:params:grant-type:ciba` |
| `auth_req_id` | Required if the `grant_type` is `urn:openid:params:grant-type:ciba`. |
| `scope` | This is a list of scopes that the client wants included in the access token. |
| `request` | If you are using `private_key_jwt` as the [client authentication method](/docs/reference/api/apps/#credentials), this is the JWT created by the client that enables you to pass requests as a single, self-contained parameter. See [Parameter details](/docs/reference/api/oidc/#client-initiated-backchannel-authentication-parameter-details). |

1. Send a request for tokens to the `/token` endpoint:

   **Example request**

   ```bash
     curl --request POST \
     --url https://${yourOktaDomain}/oauth2/default/v1/token \
     --header 'Accept: application/json' \
     --header 'authorization: Basic MG9hY...' \
     --header 'Content-Type: application/x-www-form-urlencoded' \
     --data 'grant_type=urn:openid:params:grant-type:ciba&scope=openid email&auth_req_id=ftuPCF1eZ0ujBBC3RlRPwTXucFfFQ1M3bh&login_hint=test.user@email.com'
   ```

>    > **Note:** For additional request examples using different parameters as well as a public key/private key [client authentication](/docs/reference/api/oidc/#client-authentication-methods) example, see [CIBA request examples](/docs/reference/api/oidc/#ciba-request-examples).

Reviewer 2: Again, didn't see ciba request examples. Link went to: /docs/reference/api/oidc/#possible-errors

Author: Yes, this is in Mymoona's CIBA PR https://github.com/okta/okta-developer-docs/pull/3709

as -> client: Sends authorization error
device -> as: Returns user consent/authorization
client -> as: Polls with `auth_req_id` to `/token` for tokens
as -> client: Returns requested tokens
@enduml

-->

1. The OIDC application on the consumption device triggers the CIBA authentication flow by sending a POST request to the authorization server’s backchannel authentication endpoint (`/oauth2/v1/bc/authorize`).

    To identify the user that additional authentication is being requested for, the request must include either a previously issued ID token (obtained during the user’s initial authorization) as an `id_token_hint` or the user’s username (email address) as a `login_hint`.

    > **Note:** The `id_token_hint` is used in cases where the user signed in to the application running on the consumption device and was authenticated by the Okta authorization server. The user is also in control of both the consumption and authentication device.

2. The authorization server validates the authentication request and the user identification, and then sends the backchannel authentication response with the authentication request ID (`auth_req_id`) to the client.

> 3. Okta sends a backchannel request for user consent/authorization to a branded authenticator on the user’s device (authentication device). That request contains all the information needed to authenticate the user without asking them for their credentials.

Reviewer: Can we use approval instead of consent?

4. The client (consumption device) begins to poll the `/token` endpoint including the `auth_req_id` in the request.

> 5. An authorization error is sent to the client if the authorization server hasn’t received user consent yet.

Reviewer: Can we use approval instead of consent?

    > **Note:** You can leave the rest of the default values, as they work with this guide for testing purposes.

9. In the **Assignments** section, select **Allow everyone in your organization to access**.
10. Click **Save**. The settings page for the app integration appears, showing the **General** tab.
11. Make note of the **Client ID** in the **Client Credentials** section and the client secret in the **CLIENT SECRETS** section. You need this information for the [Test the CIBA flow](#test-the-ciba-flow) section.

#### Use the API

You can also use the [Apps API](/docs/reference/api/apps/#add-oauth-2-0-client-application) to create the OAuth 2.0 client app with the CIBA grant type. Use the following parameters in the request:

**Supported parameters**

| Parameter                    | Value                                                                        |
| :--------------------------- | :--------------------------------------------------------------------------- |
| `grant_types`                | `urn:openid:params:grant-type:ciba`     |

> | `backchannel_authentication_request_signing_alg` | (Optional) The signing algorithm for CIBA signed requests using JWT. If this value isn't set and a JWT signed request is sent, the request fails. |

Reviewer: If the request fails, why is this value optional? @mymoonasaifunnissameerasahib-okta

| `backchannel_custom_authenticator_id` | The ID of the custom authenticator that authenticates the user.           |

> | `backchannel_token_delivery_mode`| (Optional) How CIBA is delivered. Valid types include `poll`, `ping`, and `push`. Since `poll` is the only value currently supported, this parameter is optional. |

Reviewer: Should we just specify `poll` if other values are not supported?

1. In the **PushSampleApp** directory of the Okta Devices SDK, run the Okta Authenticator Sample App and then open the mobile app in the simulator.
2. Click **Sign In**, enter the test user’s credentials, and click **Sign In**.
3. Complete any additional two-factor authentication required and click **Verify**.
4. On the Sign in securely using push notifications screen, click **Set up**.
5. On the Security Settings screen, slide to enable **Sign in with push notifications**.
6. Click **Ok** at the confirmation dialog. Additional security setting options then appear, including **Enable CIBA transactions**.
7. Slide to select **Enable CIBA transactions**.<br>
   The sample app is set up to include the **Enable CIBA transactions** option by default for the user to enable CIBA themselves in the mobile app. However, you can implement CIBA in your app any way that you want, for example, create an enrollment flow that turns CIBA on by default, making it transparent to users.
8. Click **Ok** at the success dialog and close the Security Settings screen.
9. Leave your test user signed in to the authentication device (the Magenta Bank app on the xCode simulator).

>    > **Note:** See the [MyAccount API](/docs/reference/api/myaccount/) for examples of [enrolling](?link to new api operation?) and [updating](?link to new api operation?) a custom app authenticator to use CIBA by adding `CIBA` to the `transactionTypes` array.

Reviewer: Assume the placeholders are intentional
